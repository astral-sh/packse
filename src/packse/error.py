import textwrap
from pathlib import Path


class PackseError(Exception):
    """Base type for all Packse exceptions"""


class UserError(PackseError):
    """It's the user's fault :)"""


class RequiresExtra(UserError):
    def __init__(self, feature: str, extra: str) -> None:
        message = f"Using {feature} requires installation with extra {extra!r}"
        super().__init__(message)


class DestinationAlreadyExists(UserError):
    def __init__(self, destination: Path) -> None:
        message = f"Destination directory '{destination}' already exists"
        super().__init__(message)


class InvalidScenario(UserError):
    def __init__(self, path: Path, reason: str) -> None:
        message = f"File at '{path}' is not a valid scenario: {reason}"
        super().__init__(message)


class InvalidPackageVersion(UserError):
    """Version is not PEP compliant"""

    def __init__(self, version: str) -> None:
        message = f"Version {version!r} is not valid."
        super().__init__(message)


class FileNotFound(UserError):
    def __init__(self, path: Path) -> None:
        message = f"File '{path}' not found"
        super().__init__(message)


class BuildError(PackseError):
    def __init__(self, message, output) -> None:
        message = message + ":\n" + textwrap.indent(output, " " * 4)
        super().__init__(message)


class PublishError(PackseError):
    pass


class PublishToolError(PublishError):
    def __init__(self, message, output) -> None:
        message = message + ":\n" + textwrap.indent(output, " " * 4)
        super().__init__(message)


class PublishNoCredentials(PublishError):
    def __init__(self) -> None:
        message = (
            "No credentials found for publish!\n"
            "Provide an API token via `PACKSE_PUBLISH_PASSWORD` or disable authentication with `--anonymous`"
        )
        super().__init__(message)


class PublishAlreadyExists(PublishError):
    def __init__(self, package: str) -> None:
        message = f"Publish for '{package}' already exists"
        super().__init__(message)


class PublishRateLimit(PublishError):
    def __init__(self, package: str) -> None:
        message = f"Publish of '{package}' failed due to rate limits"
        super().__init__(message)


class PublishConnectionError(PublishError):
    def __init__(self, package: str) -> None:
        message = f"Publish of '{package}' failed due to a connection error"
        super().__init__(message)


class InvalidPublishTarget(UserError):
    def __init__(self, path: Path, reason: str) -> None:
        message = f"Publish target at '{path}' is not a valid: {reason}"
        super().__init__(message)


class ServeError(PackseError):
    def __init__(self, message: str) -> None:
        super().__init__(message)


class ServeThreadError(ServeError):
    def __init__(self, message: str) -> None:
        super().__init__(message)


class ServeCommandError(ServeError):
    def __init__(self, message: str, stderr: str) -> None:
        if stderr.strip():
            message += f":\n\n{stderr}"
        super().__init__(message)


class ServeAddressInUse(ServeError):
    def __init__(self, url: str) -> None:
        message = f"Address '{url}' already in use"
        super().__init__(message)


class ServeAlreadyRunning(ServeError):
    def __init__(self) -> None:
        message = "Server is already running"
        super().__init__(message)
