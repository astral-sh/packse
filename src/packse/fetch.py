import logging
import shutil
import subprocess
import tempfile
import time
from pathlib import Path

import pkg_resources

from packse.error import DestinationAlreadyExists

logger = logging.getLogger(__name__)


def fetch(
    dest: Path | None = None,
    ref: str | None = None,
    repo_url: str = "https://github.com/astral-sh/packse",
    repo_subdir: str = "scenarios",
    force: bool = False,
):
    start_time = time.time()
    dest = dest or (Path.cwd() / "scenarios")

    if dest.exists() and not force:
        raise DestinationAlreadyExists(dest)

    if not ref:
        ref = pkg_resources.get_distribution("packse").version
        if ref == "0.0.0":
            ref = "HEAD"

    with tempfile.TemporaryDirectory() as tmpdir:
        # Perform a sparse checkout where we only grab the `scenarios` folder
        logger.info("Cloning repository %s", repo_url)
        subprocess.check_call(
            [
                "git",
                "clone",
                "-n",
                "--filter=tree:0",
                repo_url,
                "repo",
            ],
            stdout=subprocess.DEVNULL,
            stderr=subprocess.STDOUT,
            cwd=tmpdir,
        )

        logger.info("Checking out directory '%s' at ref %s", repo_subdir, ref)
        subprocess.check_call(
            ["git", "sparse-checkout", "set", "--no-cone", repo_subdir],
            stdout=subprocess.DEVNULL,
            stderr=subprocess.STDOUT,
            cwd=Path(tmpdir) / "repo",
        )
        subprocess.check_call(
            ["git", "fetch", "origin", ref],
            stdout=subprocess.DEVNULL,
            stderr=subprocess.STDOUT,
            cwd=Path(tmpdir) / "repo",
        )
        subprocess.check_call(
            ["git", "checkout", ref],
            stdout=subprocess.DEVNULL,
            stderr=subprocess.STDOUT,
            cwd=Path(tmpdir) / "repo",
        )

        src = Path(tmpdir) / "repo" / repo_subdir
        file_count = 0
        for file in sorted(src.iterdir()):
            file_count += 1
            logger.info("Found %s", file.name)

        logger.info("Copying files into '%s'", dest)
        shutil.copytree(src, dest, dirs_exist_ok=True)

    logger.info(
        "Fetched %s files in %.2fs",
        file_count,
        time.time() - start_time,
    )
