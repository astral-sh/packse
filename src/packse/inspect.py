"""
Get details for all scenarios.
"""
import logging
from pathlib import Path
from typing import cast

from packse.error import FileNotFound, InvalidScenario
from packse.scenario import (
    Requirement,
    Scenario,
    load_scenarios,
    scenario_hash,
)
from packse.view import dependency_tree

logger = logging.getLogger(__name__)


def inspect(
    targets: list[Path],
    skip_invalid: bool = False,
    short_names: bool = False,
    no_hash: bool = False,
):
    scenarios_by_path: dict[Path, list[Scenario]] = {}

    # Validate and collect all targets first
    for target in sorted(targets):
        if not target.exists():
            raise FileNotFound(target)

        if target.is_dir():
            for target in target.glob("*.json"):
                try:
                    logger.debug("Loading %s", target)
                    scenarios_by_path[target] = load_scenarios(target)
                except Exception as exc:
                    invalid = InvalidScenario(target, reason=str(exc))
                    if skip_invalid:
                        print(f"Skipping file: {invalid}")
                    else:
                        raise invalid
        else:
            try:
                logger.debug("Loading %s", target)
                scenarios_by_path[target] = load_scenarios(target)
            except Exception as exc:
                invalid = InvalidScenario(target, reason=str(exc))
                if skip_invalid:
                    print(f"Skipping file: {invalid}")
                else:
                    raise invalid

    # Collect a JSON-compatible representation for each scenario
    result = {"scenarios": []}
    for source, scenarios in scenarios_by_path.items():
        for scenario in scenarios:
            scenario = cast(Scenario, scenario)

            raw = scenario.dict()
            raw["source"] = str(source)
            raw["version"] = scenario_hash(scenario)
            raw["tree"] = dependency_tree(scenario).splitlines()
            result["scenarios"].append(raw)

            raw["resolver_options"]["no_build"] = [
                str(
                    Requirement(name).with_unique_name(
                        scenario,
                        raw["version"],
                        short_names,
                        no_hash=no_hash,
                    )
                )
                for name in raw["resolver_options"]["no_build"]
            ]
            raw["resolver_options"]["no_binary"] = [
                str(
                    Requirement(name).with_unique_name(
                        scenario,
                        raw["version"],
                        short_names,
                        no_hash=no_hash,
                    )
                )
                for name in raw["resolver_options"]["no_binary"]
            ]

            # Convert dictionaries to lists for easier templating
            raw["packages"] = [
                {
                    "name": str(
                        Requirement(name).with_unique_name(
                            scenario,
                            raw["version"],
                            short_names,
                            no_hash=no_hash,
                        )
                    ),
                    "versions": [
                        {
                            **package_metadata.dict(),
                            "version": version,
                            "requires": [
                                str(
                                    requirement.with_unique_name(
                                        scenario,
                                        raw["version"],
                                        short_names,
                                        no_hash=no_hash,
                                    )
                                )
                                for requirement in package_metadata.requires
                            ],
                            "extras": [
                                {
                                    "name": extra,
                                    "requires": [
                                        str(
                                            requirement.with_unique_name(
                                                scenario,
                                                raw["version"],
                                                short_names,
                                                no_hash=no_hash,
                                            )
                                        )
                                        for requirement in extra_requires
                                    ],
                                }
                                for extra, extra_requires in package_metadata.extras.items()
                            ],
                        }
                        for version, package_metadata in package.versions.items()
                    ],
                }
                for name, package in scenario.packages.items()
            ]
            raw["expected"]["packages"] = [
                {
                    "name": str(
                        Requirement(name).with_unique_name(
                            scenario,
                            raw["version"],
                            short_names,
                            no_hash=no_hash,
                        )
                    ),
                    "version": version,
                }
                for name, version in scenario.expected.packages.items()
            ]
            raw["root"]["requires"] = [
                {
                    "requirement": str(requirement),
                    "name": requirement.name,
                    "module_name": requirement.name.replace("-", "_"),
                }
                for requirement in (
                    requirement.with_unique_name(
                        scenario,
                        raw["version"],
                        short_names,
                        no_hash=no_hash,
                    )
                    for requirement in scenario.root.requires
                )
            ]

            # Ensure a module name is available for testing import of packages
            for package in raw["expected"]["packages"]:
                package["module_name"] = package["name"].replace("-", "_")

            raw["module_name"] = raw["name"].replace("-", "_")

    return result
