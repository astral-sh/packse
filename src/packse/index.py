import errno
import hashlib
import importlib
import logging
import os
import shutil
import signal
import subprocess
import sys
import tempfile
import time
from contextlib import contextmanager, nullcontext
from pathlib import Path
from typing import Generator

from packse import __development_base_path__
from packse.build import build
from packse.error import (
    RequiresExtra,
    ServeAddressInUse,
    ServeAlreadyRunning,
    ServeCommandError,
)
from packse.inspect import inspect
from packse.template import create_from_template

logger = logging.getLogger(__name__)


def index_up(
    host: str = "localhost",
    port: int = 3141,
    background: bool = False,
    dist_dir: Path | None = None,
    reset: bool = False,
    offline: bool = False,
):
    try:
        importlib.import_module("pypiserver")
    except ImportError:
        raise RequiresExtra("index commands", "index")

    if not dist_dir:
        dist_dir = Path.cwd() / "dist"
    else:
        dist_dir = Path(dist_dir).resolve()

    if background:
        if pid := get_server_pid():
            if is_running(pid):
                raise ServeAlreadyRunning()

    if reset:
        logger.info("Clearing existing server data...")
        if dist_dir.exists():
            shutil.rmtree(dist_dir)

    state_path = get_packse_state_path()

    logger.debug("Storing server metadata at %s", state_path)
    server_log_path = state_path / "server.log" if background else None
    index_url = f"http://{host}:{port}"
    index_note = "offline index" if offline else "local index with PyPI fallback"
    logger.info("Starting %s at %s", index_note, index_url)
    with start_index_server(
        dist_dir,
        host,
        port,
        server_log_path,
        offline=offline,
    ) as server_process:
        # Copy build dependencies over
        add_vendored_build_deps(dist_dir, offline)

        if background:
            logger.info("Running in background with pid %s", server_process.pid)
            logger.info("Stop index server with `packse index down`.")
            print(index_url)

        else:
            debug = logger.getEffectiveLevel() <= logging.DEBUG
            if not debug:
                logger.info("Hiding server output, use `-v` to stream server logs")

            logger.info("Ready! [Stop with Ctrl-C]")

            line = ""
            while True:
                line = server_process.stdout.readline().decode()
                if debug:
                    print(line, end="")


def index_down() -> bool:
    state_path = get_packse_state_path(create=False)

    if not state_path.exists():
        logger.warning("No server detected!")
        return False

    if pid := get_server_pid():
        if not is_running(pid):
            logger.info("Server looks shutdown already.")
            reset_pidfile()
            return

        logger.info("Stopping server with pid %s...", pid)
        try:
            os.kill(pid, signal.SIGINT)
        except ProcessLookupError:
            # Just move on if it's already dead
            pass

        # Wait for the server to exit for 10s
        limit = 10
        start_time = time.time()
        while time.time() - start_time < limit:
            if not is_running(pid):
                break
            time.sleep(0.1)
        else:
            # After the limit is reached, send a stronger signal...
            logger.info("Server has not stopped after %ds, sending kill...", limit)
            os.kill(pid, signal.SIGKILL)

            while is_running(pid):
                time.sleep(0.1)

        reset_pidfile()
        logger.info("Stopped server!")
        return True
    else:
        logger.warning("No server detected!")
        return False


@contextmanager
def start_index_server(
    dist_dir: Path,
    host: str,
    port: int,
    server_log_path: Path | None,
    offline: bool,
) -> Generator[subprocess.Popen, None, None]:
    server_url = f"http://{host}:{port}"
    dist_dir.mkdir(parents=True, exist_ok=True)

    command = [
        sys.executable,
        "-m",
        "pypiserver",
        "run",
        "--host",
        host,
        "--port",
        str(port),
        "-v",
        # Disable auth
        "-P",
        ".",
        "-a",
        ".",
        str(dist_dir),
    ]

    if offline:
        command.append("--disable-fallback")

    logger.debug("Running: %s", " ".join(command))

    server_process = subprocess.Popen(
        command,
        stderr=subprocess.STDOUT,
        stdout=server_log_path.open("wb") if server_log_path else subprocess.PIPE,
    )

    try:
        # Wait for server to be ready
        with server_log_path.open("rb") if server_log_path else nullcontext(
            server_process.stdout
        ) as server_output:
            line = ""
            lines = []

            # Watch for startup
            # TODO: Use a healthceck instead?
            while "Serving on" not in line and "Listening on" not in line:
                try:
                    server_process.wait(0)
                except subprocess.TimeoutExpired:
                    pass
                else:
                    # The process stopped already
                    if (
                        logger.getEffectiveLevel() > logging.DEBUG
                        and server_output.seekable()
                    ):
                        server_output.seek(0)
                    err_output = (
                        "\n".join(lines) + server_output.read().decode().strip()
                    )
                    raise ServeCommandError("Failed to start server!", err_output)

                line = server_output.readline().decode()
                if logger.getEffectiveLevel() <= logging.DEBUG:
                    print(line, end="")
                elif not server_output.seekable():
                    # Track the lines if we'll never be able to get them back
                    lines.append(line)
                if "Address already in use" in line:
                    raise ServeAddressInUse(server_url)

        # Server started!
        write_server_pid(server_process.pid)

        if server_log_path:
            logger.info("Writing server logs to %s", server_log_path)

        try:
            display_path = f"./{dist_dir.relative_to(Path.cwd())}"
        except Exception:
            display_path = dist_dir

        logger.info(f"Serving distributions from {display_path}")

        yield server_process

    except BaseException as exc:
        server_process.terminate()

        stdout, _ = server_process.communicate(timeout=1)
        if logger.getEffectiveLevel() <= logging.DEBUG:
            print(stdout.decode())

        server_process.kill()

        stdout, _ = server_process.communicate(timeout=1)
        if logger.getEffectiveLevel() <= logging.DEBUG:
            print(stdout.decode())

        # Do not reset the pidfile when a server was already running!
        if not isinstance(exc, ServeAddressInUse):
            reset_pidfile()

        raise


def add_vendored_build_deps(dist_dir: Path, offline: bool):
    build_directory = __development_base_path__ / "vendor" / "build"
    if build_directory.exists():
        logger.debug("Copying vendored build dependencies...")
        shutil.copytree(build_directory, dist_dir / "build", dirs_exist_ok=True)
    else:
        logger.warning("No vendored build dependencies found at %s" % build_directory)
        if offline:
            logger.error(
                "Server has no fallback PyPI access and source distribution builds are likely to fail"
            )


def get_packse_state_path(create: bool = True) -> Path:
    path = Path(os.environ.get("PACKSE_STATE_PATH") or (Path.home() / ".packse"))
    if create:
        path.mkdir(exist_ok=True)
    return path


def get_pidfile_path() -> Path:
    return get_packse_state_path() / "server.pid"


def get_server_pid() -> int | None:
    pidfile = get_pidfile_path()
    if not pidfile.exists():
        return None
    else:
        logger.debug("Found existing pid at %s", pidfile)
        return int(pidfile.read_text())


def write_server_pid(pid: int) -> None:
    pidfile = get_pidfile_path()
    pidfile.write_text(str(pid))


def reset_pidfile():
    pidfile = get_pidfile_path()
    pidfile.unlink(missing_ok=True)


def is_running(pid):
    try:
        os.kill(pid, 0)
    except OSError as err:
        # ESRCH: pid not exist
        if err.errno == errno.ESRCH:
            return False
    return True


def sha256_file(path: Path):
    h = hashlib.sha256()

    with open(path, "rb") as file:
        while True:
            # Reading is buffered, so we can read smaller chunks.
            chunk = file.read(h.block_size)
            if not chunk:
                break
            h.update(chunk)

    return h.hexdigest()


def build_index(
    targets: list[Path],
    no_hash: bool,
    short_names: bool,
    dist_dir: Path | None,
):
    out_path = Path("./index")
    if out_path.exists():
        shutil.rmtree(out_path)

    with tempfile.TemporaryDirectory() as tmpdir:
        tmpdir = Path(tmpdir)
        if not dist_dir:
            logger.info("Building distributions...")
            build(
                targets,
                rm_destination=True,
                skip_root=False,
                short_names=short_names,
                no_hash=no_hash,
                dist_dir=tmpdir / "dist",
                build_dir=tmpdir / "build",
            )
            dist_dir = tmpdir / "dist"

        out_path.mkdir()
        (out_path / "files").mkdir()

        variables = inspect(targets, short_names=short_names, no_hash=no_hash)
        for scenario in variables["scenarios"]:
            for package in scenario["packages"]:
                # Create a new distributions section
                package["dists"] = []
                for version in package["versions"]:
                    for file in Path(
                        dist_dir
                        / (
                            scenario["name"]
                            if no_hash
                            else f"{scenario['name']}-{scenario['version']}"
                        )
                    ).iterdir():
                        if package["name"].replace("-", "_") + "-" + version[
                            "version"
                        ] not in str(file):
                            continue

                        # Include all the version information to start
                        dist = version.copy()
                        # Then add a sha256
                        dist["sha256"] = sha256_file(file)
                        dist["file"] = file.name
                        package["dists"].append(dist)
                        logger.info(
                            "Found distribution %s",
                            file.name,
                        )
                        file.rename(out_path / "files" / file.name)

        logger.info("Populating template...")
        create_from_template(
            out_path,
            "index",
            variables=variables,
        )

    logger.info("Built static index at ./%s", out_path)
