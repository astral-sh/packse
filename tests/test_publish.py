import os
import re
import shutil
import stat
import subprocess
import tempfile
from pathlib import Path
from typing import Generator

import pytest
from packse import __development_base_path__

from .common import snapshot_command


@pytest.fixture(scope="module")
def scenario_dist() -> Generator[Path, None, None]:
    target = __development_base_path__ / "scenarios" / "examples" / "example.json"

    with tempfile.TemporaryDirectory() as tmpdir:
        subprocess.check_call(
            ["packse", "build", str(target)],
            cwd=tmpdir,
            stdout=subprocess.PIPE,
            stderr=subprocess.PIPE,
        )

        dists = list((Path(tmpdir) / "dist").iterdir())
        assert len(dists) == 1
        dist = dists[0]

        yield dist


@pytest.fixture
def credentials(tmpenviron) -> None:
    os.environ["PACKSE_PUBLISH_USERNAME"] = "username"
    os.environ["PACKSE_PUBLISH_PASSWORD"] = "password"
    yield


class MockBinary:
    def __init__(self, path: Path) -> None:
        self.path = path
        self.callback = None
        self._update_bin("")

    def _prepare_text(self, text: str = None):
        if not text:
            return ""
        # Escape single quotes
        return text.replace("'", "\\'")

    def _update_bin(self, content: str):
        self.path.write_text("#!/usr/bin/env sh\n\n" + content + "\n")
        self.path.chmod(self.path.stat().st_mode | stat.S_IEXEC)

    def set_success(self, text: str | None = None):
        text = self._prepare_text(text)
        self._update_bin(f"echo '{text}'")

    def set_error(self, text: str | None = None):
        text = self._prepare_text(text)
        self._update_bin(f"echo '{text}'; exit 1")


@pytest.fixture
def mock_twine(monkeypatch: pytest.MonkeyPatch) -> Generator[MockBinary, None, None]:
    # Create a temp directory to register as a bin
    with tempfile.TemporaryDirectory() as tmpdir:
        mock = MockBinary(Path(tmpdir) / "twine")
        mock.set_success("<mock twine logs>")

        # Add to the path
        monkeypatch.setenv("PATH", tmpdir, prepend=":")
        assert shutil.which("twine").startswith(tmpdir)

        yield mock


@pytest.mark.usefixtures("credentials")
def test_publish_example_dry_run(snapshot, scenario_dist: Path):
    assert (
        snapshot_command(
            ["publish", "--dry-run", scenario_dist],
            extra_filters=[(re.escape(str(scenario_dist.resolve())), "[DISTDIR]")],
        )
        == snapshot
    )


@pytest.mark.usefixtures("credentials")
def test_publish_example_twine_succeeds(
    snapshot, scenario_dist: Path, mock_twine: MockBinary
):
    mock_twine.set_success("<twine happy message>")

    assert (
        snapshot_command(
            ["publish", scenario_dist, "-v", "--workers", "1"],
            extra_filters=[(re.escape(str(scenario_dist.resolve())), "[DISTDIR]")],
        )
        == snapshot
    )


@pytest.mark.usefixtures("credentials")
def test_publish_example_twine_succeeds_parallel(
    snapshot, scenario_dist: Path, mock_twine: MockBinary
):
    mock_twine.set_success("<twine happy message>")

    assert (
        snapshot_command(
            ["publish", scenario_dist, "-v"],
            extra_filters=[(re.escape(str(scenario_dist.resolve())), "[DISTDIR]")],
            # Cannot record stderr when running in parallel
            snapshot_stderr=False,
        )
        == snapshot
    )


@pytest.mark.usefixtures("credentials")
def test_publish_example_twine_fails_with_unknown_error(
    snapshot, scenario_dist: Path, mock_twine: MockBinary
):
    mock_twine.set_error("<twine error message>")

    assert (
        snapshot_command(
            ["publish", scenario_dist, "-v", "--workers", "1"],
            extra_filters=[(re.escape(str(scenario_dist.resolve())), "[DISTDIR]")],
        )
        == snapshot
    )


@pytest.mark.usefixtures("credentials")
def test_publish_example_twine_fails_with_rate_limit(
    snapshot, scenario_dist: Path, mock_twine: MockBinary
):
    mock_twine.set_error(
        """
Uploading distributions to https://test.pypi.org/legacy/
Uploading
requires_transitive_incompatible_with_root_version_5c1b7dc1_c-1.0.0-py3-none-any
.whl
25l
    0% ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 0.0/4.1 kB • --:-- • ?
100% ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 4.1/4.1 kB • 00:00 • ?
100% ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 4.1/4.1 kB • 00:00 • ?
100% ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 4.1/4.1 kB • 00:00 • ?
100% ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 4.1/4.1 kB • 00:00 • ?
25hWARNING  Error during upload. Retry with the --verbose option for more details.
ERROR    HTTPError: 429 Too Many Requests from https://test.pypi.org/legacy/
            Too many new projects created
        """
    )

    assert (
        snapshot_command(
            ["publish", scenario_dist, "--workers", "1"],
            extra_filters=[(re.escape(str(scenario_dist.resolve())), "[DISTDIR]")],
        )
        == snapshot
    )


@pytest.mark.usefixtures("credentials")
def test_publish_example_twine_fails_with_already_exists(
    snapshot, scenario_dist: Path, mock_twine: MockBinary
):
    mock_twine.set_error(
        """
Uploading distributions to https://test.pypi.org/legacy/
Uploading example_9e723676_a-1.0.0.tar.gz
100% ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 3.1/3.1 kB • 00:00 • ?
WARNING  Error during upload. Retry with the --verbose option for more details.
ERROR    HTTPError: 400 Bad Request from https://test.pypi.org/legacy/
         File already exists. See https://test.pypi.org/help/#file-name-reuse for more information.
        """
    )

    assert (
        snapshot_command(
            ["publish", scenario_dist, "-v", "--workers", "1"],
            extra_filters=[(re.escape(str(scenario_dist.resolve())), "[DISTDIR]")],
        )
        == snapshot
    )


@pytest.mark.usefixtures("tmpenviron")
def test_publish_example_no_credentials(
    snapshot,
    scenario_dist: Path,
    mock_twine: MockBinary,
):
    # Ensure these do not exist
    if "PACKSE_PUBLISH_PASSWORD" in os.environ:
        os.environ.pop("PACKSE_PUBLISH_PASSWORD")
    if "TWINE_PASSWORD" in os.environ:
        os.environ.pop("TWINE_PASSWORD")

    mock_twine.set_error(
        """
        Should not be reached!
        """
    )

    assert (
        snapshot_command(
            ["publish", scenario_dist, "-v", "--workers", "1"],
            extra_filters=[(re.escape(str(scenario_dist.resolve())), "[DISTDIR]")],
        )
        == snapshot
    )


@pytest.mark.usefixtures("credentials")
def test_publish_example_no_username_defaults_to_token(
    snapshot,
    scenario_dist: Path,
    mock_twine: MockBinary,
):
    # Ensure these do not exist
    if "PACKSE_PUBLISH_USERNAME" in os.environ:
        os.environ.pop("PACKSE_PUBLISH_USERNAME")
    if "TWINE_USERNAME" in os.environ:
        os.environ.pop("TWINE_USERNAME")

    assert (
        snapshot_command(
            ["publish", scenario_dist, "-v", "--workers", "1"],
            extra_filters=[(re.escape(str(scenario_dist.resolve())), "[DISTDIR]")],
        )
        == snapshot
    )
