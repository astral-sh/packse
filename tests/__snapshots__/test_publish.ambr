# serializer version: 1
# name: test_publish_example_dry_run
  dict({
    'exit_code': 0,
    'stderr': '''
      Publishing 1 target to https://test.pypi.org/legacy/...
      Publishing 'example-961b4c22'...
      Published 'example_961b4c22-0.0.0.tar.gz'
      Published 'example_a_961b4c22-1.0.0-py3-none-any.whl'
      Published 'example_a_961b4c22-1.0.0.tar.gz'
      Published 'example_b_961b4c22-1.0.0-py3-none-any.whl'
      Published 'example_b_961b4c22-1.0.0.tar.gz'
      Published 'example_b_961b4c22-2.0.0-py3-none-any.whl'
      Published 'example_b_961b4c22-2.0.0.tar.gz'
      Published 'example_b_961b4c22-3.0.0-py3-none-any.whl'
      Published 'example_b_961b4c22-3.0.0.tar.gz'
      Published 1 targets (9 new files) in [TIME]
  
    ''',
    'stdout': '''
      Would execute: twine upload --repository-url https://test.pypi.org/legacy/ [DISTDIR]/example_961b4c22-0.0.0.tar.gz
      Would execute: twine upload --repository-url https://test.pypi.org/legacy/ [DISTDIR]/example_a_961b4c22-1.0.0-py3-none-any.whl
      Would execute: twine upload --repository-url https://test.pypi.org/legacy/ [DISTDIR]/example_a_961b4c22-1.0.0.tar.gz
      Would execute: twine upload --repository-url https://test.pypi.org/legacy/ [DISTDIR]/example_b_961b4c22-1.0.0-py3-none-any.whl
      Would execute: twine upload --repository-url https://test.pypi.org/legacy/ [DISTDIR]/example_b_961b4c22-1.0.0.tar.gz
      Would execute: twine upload --repository-url https://test.pypi.org/legacy/ [DISTDIR]/example_b_961b4c22-2.0.0-py3-none-any.whl
      Would execute: twine upload --repository-url https://test.pypi.org/legacy/ [DISTDIR]/example_b_961b4c22-2.0.0.tar.gz
      Would execute: twine upload --repository-url https://test.pypi.org/legacy/ [DISTDIR]/example_b_961b4c22-3.0.0-py3-none-any.whl
      Would execute: twine upload --repository-url https://test.pypi.org/legacy/ [DISTDIR]/example_b_961b4c22-3.0.0.tar.gz
      example-961b4c22
  
    ''',
  })
# ---
# name: test_publish_example_no_credentials
  dict({
    'exit_code': 1,
    'stderr': '''
      No credentials found for publish!
      Provide an API token via `PACKSE_PUBLISH_PASSWORD` or disable authentication with `--anonymous`.
  
    ''',
    'stdout': '',
  })
# ---
# name: test_publish_example_no_username_defaults_to_token
  dict({
    'exit_code': 0,
    'stderr': '''
      Publishing 1 target to https://test.pypi.org/legacy/...
      Publishing 'example-961b4c22'...
      Published example_961b4c22-0.0.0.tar.gz in [TIME]:
      
          <mock twine logs>
      
      Published 'example_961b4c22-0.0.0.tar.gz'
      Published example_a_961b4c22-1.0.0-py3-none-any.whl in [TIME]:
      
          <mock twine logs>
      
      Published 'example_a_961b4c22-1.0.0-py3-none-any.whl'
      Published example_a_961b4c22-1.0.0.tar.gz in [TIME]:
      
          <mock twine logs>
      
      Published 'example_a_961b4c22-1.0.0.tar.gz'
      Published example_b_961b4c22-1.0.0-py3-none-any.whl in [TIME]:
      
          <mock twine logs>
      
      Published 'example_b_961b4c22-1.0.0-py3-none-any.whl'
      Published example_b_961b4c22-1.0.0.tar.gz in [TIME]:
      
          <mock twine logs>
      
      Published 'example_b_961b4c22-1.0.0.tar.gz'
      Published example_b_961b4c22-2.0.0-py3-none-any.whl in [TIME]:
      
          <mock twine logs>
      
      Published 'example_b_961b4c22-2.0.0-py3-none-any.whl'
      Published example_b_961b4c22-2.0.0.tar.gz in [TIME]:
      
          <mock twine logs>
      
      Published 'example_b_961b4c22-2.0.0.tar.gz'
      Published example_b_961b4c22-3.0.0-py3-none-any.whl in [TIME]:
      
          <mock twine logs>
      
      Published 'example_b_961b4c22-3.0.0-py3-none-any.whl'
      Published example_b_961b4c22-3.0.0.tar.gz in [TIME]:
      
          <mock twine logs>
      
      Published 'example_b_961b4c22-3.0.0.tar.gz'
      Published 1 targets (9 new files) in [TIME]
  
    ''',
    'stdout': '''
      example-961b4c22
  
    ''',
  })
# ---
# name: test_publish_example_twine_fails_with_already_exists
  dict({
    'exit_code': 1,
    'stderr': '''
      Publishing 1 target to https://test.pypi.org/legacy/...
      Publishing 'example-961b4c22'...
      Publish for 'example_961b4c22-0.0.0.tar.gz' already exists.
      Failed to publish 1/1 targets.
  
    ''',
    'stdout': '',
  })
# ---
# name: test_publish_example_twine_fails_with_rate_limit
  dict({
    'exit_code': 1,
    'stderr': '''
      Publishing 1 target to https://test.pypi.org/legacy/...
      Publishing 'example-961b4c22'...
      Publish of 'example_961b4c22-0.0.0.tar.gz' failed due to rate limits.
      Failed to publish 1/1 targets.
  
    ''',
    'stdout': '',
  })
# ---
# name: test_publish_example_twine_fails_with_unknown_error
  dict({
    'exit_code': 1,
    'stderr': '''
      Publishing 1 target to https://test.pypi.org/legacy/...
      Publishing 'example-961b4c22'...
      Publishing example_961b4c22-0.0.0.tar.gz with twine failed:
          <twine error message>
      .
      Failed to publish 1/1 targets.
  
    ''',
    'stdout': '',
  })
# ---
# name: test_publish_example_twine_succeeds
  dict({
    'exit_code': 0,
    'stderr': '''
      Publishing 1 target to https://test.pypi.org/legacy/...
      Publishing 'example-961b4c22'...
      Published example_961b4c22-0.0.0.tar.gz in [TIME]:
      
          <twine happy message>
      
      Published 'example_961b4c22-0.0.0.tar.gz'
      Published example_a_961b4c22-1.0.0-py3-none-any.whl in [TIME]:
      
          <twine happy message>
      
      Published 'example_a_961b4c22-1.0.0-py3-none-any.whl'
      Published example_a_961b4c22-1.0.0.tar.gz in [TIME]:
      
          <twine happy message>
      
      Published 'example_a_961b4c22-1.0.0.tar.gz'
      Published example_b_961b4c22-1.0.0-py3-none-any.whl in [TIME]:
      
          <twine happy message>
      
      Published 'example_b_961b4c22-1.0.0-py3-none-any.whl'
      Published example_b_961b4c22-1.0.0.tar.gz in [TIME]:
      
          <twine happy message>
      
      Published 'example_b_961b4c22-1.0.0.tar.gz'
      Published example_b_961b4c22-2.0.0-py3-none-any.whl in [TIME]:
      
          <twine happy message>
      
      Published 'example_b_961b4c22-2.0.0-py3-none-any.whl'
      Published example_b_961b4c22-2.0.0.tar.gz in [TIME]:
      
          <twine happy message>
      
      Published 'example_b_961b4c22-2.0.0.tar.gz'
      Published example_b_961b4c22-3.0.0-py3-none-any.whl in [TIME]:
      
          <twine happy message>
      
      Published 'example_b_961b4c22-3.0.0-py3-none-any.whl'
      Published example_b_961b4c22-3.0.0.tar.gz in [TIME]:
      
          <twine happy message>
      
      Published 'example_b_961b4c22-3.0.0.tar.gz'
      Published 1 targets (9 new files) in [TIME]
  
    ''',
    'stdout': '''
      example-961b4c22
  
    ''',
  })
# ---
# name: test_publish_example_twine_succeeds_parallel
  dict({
    'exit_code': 0,
    'stderr': '<not included>',
    'stdout': '''
      example-961b4c22
  
    ''',
  })
# ---
